Introduce `test-integration` target (and deprecate/freeze `test-integration-cli`)

vdemeester opened the pull request:
This adds a new package `integration` where `moby` integration tests
should live. Those integration tests should not depends on any `cli`
components (except from the `dockerd` daemon for now — to actually
start a daemon).

This is a simple, *naive* approach and is a *middle ground* between current `integration-cli` and `unit test`.

The *related* topics is [Evolution of testing @moby](https://forums.mobyproject.org/t/evolution-of-testing-moby/38/8), issue #32866.

- [ ] setup/cleanup *defer* should be extracted in another package
- [ ] same fo `createClient` 
- [x] should *restore* the `protectXX` too
- [ ] `integration-cli/environment` and `/integration-cli/fixtures/load` should move to `integration` package
- [ ] remove `docker` binary dependency from `integration-cli/environment` (the cleanup is using `docker` commands for now)

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

mlaventure commented on integration/container/create_test.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	// TODO: ensure and protect images
+	// 2. Run the test(s)
```
nit: can be removed

vdemeester replied:
:stuck_out_tongue:

mlaventure commented on hack/make/.integration-test-helpers:
```diff
@@ -7,6 +7,25 @@ bundle_test_integration_cli() {
 }
 
+bundle_test_integration() {
+	TESTFLAGS="$TESTFLAGS -v -test.timeout=60m"
```
maybe the reverse? if I specify `-test.timeout` in my `TESTFLAGS` it'll be overridden by the script if I'm not mistaken.

vdemeester replied:
@mlaventure hum right :angel: I just copied the others `bundle_xx` but yep, make sens

aaronlehmann commented on integration/container/create_test.go:
```diff
@@ -0,0 +1,140 @@
+
+	for index, tc := range testCases {
+		tc := tc
+		t.Run(fmt.Sprintf("%d", index), func(t *testing.T) {
```
`strconv.FormatInt`

mlaventure reviewed (changes requested):
couple of nits

mlaventure commented:
```
22:23:52 ---> Making bundle: test-unit (in bundles/17.06.0-dev/test-unit)
22:23:52 /go/src/github.com/docker/docker/hack/make/test-unit: line 40: syntax error near unexpected token `|'
```

Janky failure

aaronlehmann commented on integration/container/create_test.go:
```diff
@@ -0,0 +1,140 @@
+
+	for index, tc := range testCases {
+		tc := tc
+		t.Run(fmt.Sprintf("%d", index), func(t *testing.T) {
```
`strconv.FormatInt`

aaronlehmann commented on integration/container/create_test.go:
```diff
@@ -0,0 +1,140 @@
+		tc := tc
+		t.Run(fmt.Sprintf("%d", index), func(t *testing.T) {
+			t.Parallel()
+			_, err := clt.ContainerCreate(context.TODO(),
```
Do you expect to plumb context through the test cases at some point? Otherwise, I think `TODO` is not the right context to use here (we should probably have `ctx := context.Background()` above.

vdemeester replied:
right :)

aaronlehmann commented:
I'm wondering if it would be simpler to rename `integration-cli` to `integration` and gradually migrate existing tests to use the API directly, and/or move them elswhere.

It would be nice to avoid having two packages/targets with similar names that run different sets of tests. I could see confusion coming up between the two, such as people accidentally adding tests to `integration-cli`, or running only one of the two to validate changes.

vdemeester commented:
@aaronlehmann 

> I'm wondering if it would be simpler to rename integration-cli to integration and gradually migrate existing tests to use the API directly, and/or move them elsewhere.

Right, that's more or less what I was going for in https://github.com/moby/moby/pull/33124. The thing is, rewriting them (and/or extracting them) will take a long time, and depending on where we go with those new tests (i.e. using `testing` like this PR, or keep using `go-check`) the way we should run them could be really different.

> It would be nice to avoid having two packages/targets with similar names that run different sets of tests. I could see confusion coming up between the two, such as people accidentally adding tests to integration-cli, or running only one of the two to validate changes.

Yeah I can see confusing coming up too. I'm still not 100% sold on having those two separate either, it's just a proposal 👼. 

What I would like to be discussed and decided on for this PR are the following :

- [ ] `test-integration` & `test-integration-cli` or have only a `test-integration` package with all of them, but some freezed and being worked on to move elsewhere or rewrote ?
- [ ] `testing` (like the current proposal, with the use of subtest, etc..) or `go-check` ?

From there, this PR will be the start of the work and examples 😉

vdemeester committed:
Introduce `test-integration` target (and deprecate/freeze `test-integration-cli`)

This adds a new package `integration` where `moby` integration tests
should live. Those integration tests should not depends on any `cli`
components (except from the `dockerd` daemon for now — to actually
start a daemon).

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

aaronlehmann commented:
> test-integration & test-integration-cli or have only a test-integration package with all of them, but some freezed and being worked on to move elsewhere or rewrote ?

I prefer to have a single package during the transition.

dnephin commented:
+1 for `testing` , go-check doesn't seem to provide much value. 

A single package is a problem. It makes running a subset of the tests much slower than it needs to be.  We should really have a package for each test suite, and a suite for each component.

A single target for CI SGTM.

aaronlehmann commented:
> A single package is a problem. It makes running a subset of the tests much slower than it needs to be. We should really have a package for each test suite, and a suite for each component.
>
> A single target for CI SGTM.

That also sounds good to me. What I meant was that I'd like to have a single CI target.

vdemeester commented:
> A single package is a problem. It makes running a subset of the tests much slower than it needs to be. We should really have a package for each test suite, and a suite for each component.

@dnephin Either way, it will be splitted into several packages for sure. *A single package* is just for the *main* package, i.e. `integration` or `integration-cli` (i.e. any new tests written should target the `api` **and** be on sub package of `integration`).

thaJeztah commented on Makefile:
```diff
@@ -158,8 +158,12 @@ test: build ## run the unit, integration and docker-py tests
 test-docker-py: build ## run the docker-py tests
 	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary test-docker-py
 
+# Deprecated: use `test-integration` instead
```
Can you add the "deprecation" message in the comment below? the `## run the integration tests` comment is also used for outputting the help message (`make help`)

thaJeztah commented:
@dnephin @aaronlehmann what's the status on this one?

vdemeester commented:
@thaJeztah I need to re-work that one 👼

vdemeester commented:
Closing this temporarly 👼
